Mathlib/CategoryTheory/Abelian/GrothendieckAxioms.lean

```diff
@@ -54,58 +56,87 @@ universe v v' u u' w
 
 variable (C : Type u) [Category.{v} C]
 
-attribute [local instance] hasCoproducts_of_finite_and_filtered
+/--
+A category `C` is said to have exact colimits of shape `J` provided that colimits of shape `J`
+exist and are exact (in the sense that they preserve finite limits).
+-/
+class HasExactColimitsOfShape (J : Type u') [Category.{v'} J] (C : Type u) [Category.{v} C]
+    [HasColimitsOfShape J C]  where
+  /-- Exactness of `J`-shaped colimits stated as `colim : (J ⥤ C) ⥤ C` preserving finite limits. -/
+  preservesFiniteLimits : PreservesFiniteLimits (colim (J := J) (C := C))
+
+/--
+A category `C` is said to have exact limits of shape `J` provided that limits of shape `J`
+exist and are exact (in the sense that they preserve finite colimits).
+-/
+class HasExactLimitsOfShape (J : Type u') [Category.{v'} J] (C : Type u) [Category.{v} C]
+    [HasLimitsOfShape J C] where
+  /-- Exactness of `J`-shaped limits stated as `lim : (J ⥤ C) ⥤ C` preserving finite colimits. -/
+  preservesFiniteColimits : PreservesFiniteColimits (lim (J := J) (C := C))
+
+attribute [instance] HasExactColimitsOfShape.preservesFiniteLimits
+  HasExactLimitsOfShape.preservesFiniteColimits
 
 /--
 A category `C` which has coproducts is said to have `AB4` provided that
 coproducts are exact.
 -/
 class AB4 [HasCoproducts C] where
```

Comment:
For some future applications (e.g. étale sheaves), we shall need a more general `AB4` notion. Could you introduce a class `AB4OfSize.{w}` (with `w` a priori unrelated to `u` and `v`) which would assume coproducts of size `w` and require the same `PreservesFiniteLimits` but for all `α : Type w`, and then `AB4` would be an abbrev for `AB4OfSize.{v}`. The same remark applies to `AB5`, etc.

Reply:
I have done this and added some `shrink` API as well